clients/src/main/java/org/apache/kafka/clients/consumer/internals/WakeupTrigger.java

```diff
@@ -0,0 +1,105 @@
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.WakeupException;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Ensures blocking APIs can be woken up by the consumer.wakeup().
+ */
+public class WakeupTrigger {
+    private AtomicReference<Wakeupable> pendingTask = new AtomicReference<>(null);
+
+    /*
+      Wakeup a pending task.  If there isn't any pending task, return a WakedupFuture, so that the subsequent call
+      would know wakeup was previously called.
+
+      If there are active tasks, complete it with WakeupException, then unset pending task (return null here.
+      If the current task has already been woken-up, do nothing.
+     */
+    public void wakeup() {
+        pendingTask.getAndUpdate(task -> {
+            if (task == null) {
+                return new WakedupFuture();
+            } else if (task instanceof ActiveFuture) {
+                ActiveFuture active = (ActiveFuture) task;
+                active.future().completeExceptionally(new WakeupException());
+                return null;
+            } else {
+                return task;
+            }
+        });
+    }
+
+    /*
+    If there is no pending task, set the pending task active.
+    If wakeup was called before setting an active task, the current task will complete exceptionally with
+    WakeupException right
+    away.
+    if there is an active task, throw exception.
+     */
+    public <T> CompletableFuture<T> setActiveTask(final CompletableFuture<T> currentTask) {
+        Objects.requireNonNull(currentTask, "currentTask cannot be null");
+        pendingTask.getAndUpdate(task -> {
+            if (task == null) {
+                return new ActiveFuture(currentTask);
+            } else if (task instanceof WakedupFuture) {
+                currentTask.completeExceptionally(new WakeupException());
+                return null;
+            }
+            // last active state is still active
+            throw new KafkaException("Last active task is still active");
+        });
+        return currentTask;
+    }
+
+    public void clearActiveTask() {
+        pendingTask.getAndUpdate(task -> {
+            if (task == null) {
+                return null;
+            } else if (task instanceof ActiveFuture) {
+                return null;
+            }
+            return task;
+        });
+    }
+
+    Wakeupable getPendingTask() {
+        return pendingTask.get();
+    }
+
+    interface Wakeupable { }
+
+    static class ActiveFuture implements Wakeupable {
+        private final CompletableFuture<?> future;
+
+        public ActiveFuture(final CompletableFuture<?> future) {
+            this.future = future;
+        }
+
+        public CompletableFuture<?> future() {
+            return future;
+        }
+    }
+
+    static class WakedupFuture implements Wakeupable { }
```

Comment:
Should this be WokenUpFuture?

Reply:
Maybe just WakeupFuture? I guess it is a future for waking up a call.